File: 002_AgentesDeResolucaoDeProblemas/src/Busca.java
import java.util.Queue;
import java.util.Vector;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

public class Busca {
    private Queue<Estado> borda = new LinkedList<Estado>();
    private Vector<Estado> explorados = new Vector<Estado>();

    // RETORNA UM NÓ, O CAMINHO SÃO OS SEUS PAIS
    /*public No buscaEmLargura(Problema problema) {
        this.borda.add(problema.estadoInicial);

        while (true) {
            if (this.borda.isEmpty()) {
                return null;
            }

            No no = new No(this.borda.poll(), null, 0);
            this.explorados.add(no.estado);

            Map<Estado, Estado[]> transicoes = new HashMap<>(problema.transicoes);

            for (int i = 0; i < problema.transicoes.get(no.estado).length; i++) {
                No filho = new No(problema.transicoes.get(no.estado)[i], no, no.custoCaminho + 1);

                if (this.borda.contains(filho.estado) == false && this.explorados.contains(filho.estado) == false) {
                    if (filho.estado == problema.objetivo) {
                        return filho;
                    }
                    this.borda.add(filho.estado);
                }
            }
        }
    }*/

    // LINHAS QUE MUDAM 6, 13, 14
    public No buscaCustoUniforme(Problema problema) {
        borda.add(problema.estadoInicial);

        while (true) {
            if (borda.isEmpty()) {
                return null;
            }

            No no = new No(borda.poll(), null, 0);

            if (no.estado == problema.objetivo) {
                return no;
            }

            explorados.add(no.estado);

            for (int i = 0; i < problema.transicoes.get(no.estado).length; i++) {
                No filho = new No(problema.transicoes.get(no.estado)[i], no, no.custoCaminho);
                if (this.borda.contains(filho.estado) == false && this.explorados.contains(filho.estado) == false) {
                    borda.add(filho.estado);
                } else if () {

                }
            }
        }
    }
}

File: 002_AgentesDeResolucaoDeProblemas/Busca.py
from time import sleep
from Problema import Problema
from No import No
import heapq

class Busca:
    def __init__(self):
        self.borda = []         # guarda os nós que ainda não foram explorados
        self.explorados = []    # guarda os nós que já foram explorados

    def findPesoEstado(self, lista, estado):
        for i in range(len(lista)):
            if lista[i][0] == estado:
                return i

    def buscaEmLargura(self, problema: Problema):
        self.borda.append(problema.estadoInicial)
        root = No(problema.estadoInicial, None, 0)

        while True:
            if len(self.borda) == 0:
                return None

            a = self.borda.pop()
            indice = self.findPesoEstado(problema.transicoes[root.estado], a)

            no = No(a, root, root.custo + (0 if (indice == None) else problema.transicoes[root.estado][indice][1]))
            self.explorados.append(no)

            for child in problema.transicoes[no.estado]:
                filho = No(child[0], no, no.custo + child[1])
                # print(filho.estado)

                if (filho.estado not in self.explorados) and (filho not in self.borda):
                    if filho.estado == problema.estadoFinal:
                        return filho
                    self.borda.append(filho.estado)
                    root = filho.pai